### DISCUSSION TITLE
How to submit an example project?

### DISCUSSION THREAD

**Discussion by ROSW6341** [Bodmer/TFT_eSPI#1646] [Q&A - General] [answered]
I've updated the example 'graph' project with the following modifications & extensions:
  - Renamed to 'Chart'
  - Separated Graph & Traces
  - Added optional right side Y axis
  - C++ class
  - Suggest using TFT_eSPI version 2.4.22 or later w/ updated drawFloat()
  - Suggest copying these files to the project directory to facilitate 'tweaking'
    on a per project basis

Not familiar with submitting to github.  Is a RAR file acceptable & if so, to whom do I send it.
- Bill K <billk@rosw.com>

**Comment by Bodmer** [comment, marked as answer]
The simplest way to provide a new example sketch is to zip it and drag-drop here in a message box.

To create a sketch and maintain it (respond to user issues and bug reports) then creating your own repository is the way to go, in that case I can provide a link in the ReadMe.

**Comment by ROSW6341** [comment]
Thanks for the info.  A ZIP of the sketch is attached.  There is probably more that could be done to the code but so far it works for me.  Hope it is useful to others.
[ChartTest-20220217A.zip](https://github.com/Bodmer/TFT_eSPI/files/8098315/ChartTest-20220217A.zip)
PS Feel free to have a look and clean up any obvious &/or blatant errors.
.

**Reply by Bodmer** [reply]
I loaded and ran it and it is working very well.

Ideally it needs to use the display width and height so the demo automatically fits any user screen.

**Reply by ROSW6341** [reply]
Glad it worked OK for you.  Feel free to edit the code to use the display width & height.
Have done a bit more work on the chart code:
 -   changed all doubles to floats
 -  made dp's unsigned (uint8_t)
 -  changed the map-like code to call a common function that takes floats for the input & input range values but now takes longs for the output range values and returns a long.
 -  all location values sent to TFT_eSPI is no longer doubles but are now ints

Can merge the changes back into the example (& change the display width & height) and send you an update if you want.

Came across something that might be of interest to you.  It is a simple means for rounding floats/doubles based upon the value of dp. (sorry, messed up the original version of the following)

  float round(float value, int dp) {
    float tens = 10 ^ dp;     // -100, -10, 0, 10, 100, ...
    return round(value * tens) / tens;
  }

Might be better to make 'tens' an int or unsigned, check the sign of 'dp', and swap the mult & div (*, /) sequence accordingly.
Should to work properly for both positive & negative values but has not been tested.

**Reply by ROSW6341** [reply]
Did some more checking.  10^dp doesn't work properly at least for the esp32.  10^1 returns 11, not 10 for the Arduino ESP32 tools I'm using.  Reworked the code and did a fair amount of testing (+/- values of dp and of value).  The function as it now stands is:

float round(float value, int dp) {
  if (dp == 0)
    return round(value);

  int tens = 1;

  // calculate power of 10 from abs(dp)
  // 10, 100, 1000, ...
  for (int i = 0; i < abs(dp); i++)
    tens *= 10;

  if (dp < 0)
    return round(value / tens) * tens;

  return round(value * tens) / tens;
}
